network/stub/network.go

```diff
@@ -258,33 +232,41 @@ func (n *Network) sendToAllTargets(m *PendingMessage, syncOnProcess bool) error
 			continue
 		}
 
-		// checks if the given engine already received the event.
-		// this prevents a node receiving the same event twice.
-		if receiverNetwork.haveSeen(key) {
-			continue
+		// finds the engine of the targeted Network
+		err := receiverNetwork.processWithEngine(syncOnProcess, key, m)
+		if err != nil {
+			return fmt.Errorf("could not process message for nodeID: %v, %w", nodeID, err)
 		}
+	}
+	return nil
+}
 
-		// marks the peer has seen the event
-		receiverNetwork.seen(key)
+func (n *Network) processWithEngine(syncOnProcess bool, key string, m *PendingMessage) error {
+	n.Lock()
+	defer n.Unlock()
 
-		// finds the engine of the targeted Network
-		receiverEngine, ok := receiverNetwork.engines[m.Channel]
-		if !ok {
-			return fmt.Errorf("could find engine ID: %v for node: %v", m.Channel, nodeID)
-		}
+	// checks if the given engine already received the event.
+	// this prevents a node receiving the same event twice.
+	if _, ok := n.seenEventIDs[key]; ok {
+		return nil
+	}
+	n.seenEventIDs[key] = struct{}{}
 
-		if syncOnProcess {
-			// sender and receiver are synced over processing the message
-			if err := receiverEngine.Process(m.Channel, m.From, m.Event); err != nil {
-				return fmt.Errorf("receiver engine failed to process event (%v): %w", m.Event, err)
-			}
-		} else {
-			// sender and receiver are synced over delivery of message
-			go func() {
-				_ = receiverEngine.Process(m.Channel, m.From, m.Event)
-			}()
-		}
+	receiverEngine, ok := n.engines[m.Channel]
+	if !ok {
+		return fmt.Errorf("could find engine ID: %v", m.Channel)
```

Comment:
should still mention which node didn't have an engine?

Reply:
it is added up the stack since this error will be wrapped with:
```
		err := receiverNetwork.processWithEngine(syncOnProcess, key, m)
		if err != nil {
			return fmt.Errorf("could not process message for nodeID: %v, %w", nodeID, err)
		}
```